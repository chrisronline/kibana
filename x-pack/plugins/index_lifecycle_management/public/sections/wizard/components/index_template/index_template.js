/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { toastNotifications } from 'ui/notify';

import { TemplateSelection } from './components/template_selection';
import { Configuration } from './components/configuration';

import {
  EuiSpacer,
  EuiTitle,
  EuiHorizontalRule,
  EuiButton,
  EuiTextColor,
  EuiLink,
} from '@elastic/eui';
import { hasErrors } from '../../../../lib/find_errors';
import {
  STRUCTURE_TEMPLATE_SELECTION,
  STRUCTURE_CONFIGURATION,
} from '../../../../store/constants';

export class IndexTemplate extends Component {
  static propTypes = {
    done: PropTypes.func.isRequired,

    errors: PropTypes.object,
  };

  constructor(props) {
    super(props);
    this.state = {
      isShowingErrors: false,
    };
  }

  validate = async () => {
    await this.props.validate();
    const noErrors = !hasErrors(this.props.errors);
    return noErrors;
  };

  submit = async () => {
    this.setState({ isShowingErrors: true });
    if (await this.validate()) {
      this.props.done();
    } else {
      toastNotifications.addDanger('Please fix errors on the page.');
    }
  };

  render() {
    const { errors } = this.props;
    const { isShowingErrors } = this.state;

    return (
      <div className="euiAnimateContentLoad">
        <EuiTitle>
          <h4>Select a template</h4>
        </EuiTitle>
        <EuiTitle size="xs">
          <Fragment>
            <EuiSpacer size="xs"/>
            <EuiTextColor color="subdued">
              <h5>
                An index template defines the settings, mappings, and aliases to apply
                when you create an index.
              </h5>
            </EuiTextColor>
            <EuiSpacer size="xs"/>
            <EuiLink href="https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html">
              Learn more
            </EuiLink>
          </Fragment>
        </EuiTitle>
        <EuiSpacer />
        <TemplateSelection
          validate={this.validate}
          errors={errors[STRUCTURE_TEMPLATE_SELECTION]}
          isShowingErrors={isShowingErrors}
        />
        <Configuration
          validate={this.validate}
          errors={errors[STRUCTURE_CONFIGURATION]}
          isShowingErrors={isShowingErrors}
        />
        <EuiHorizontalRule className="ilmHrule" />

        <EuiButton
          fill
          iconSide="right"
          iconType="sortRight"
          onClick={this.submit}
        >
          Next
        </EuiButton>
      </div>
    );
  }
}
