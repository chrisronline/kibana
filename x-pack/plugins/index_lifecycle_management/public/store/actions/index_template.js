/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */




import { createAction } from 'redux-actions';
import { toastNotifications } from 'ui/notify';
import { loadIndexTemplates, loadIndexTemplate } from '../../api';
import {
  setIndexName,
  setAliasName,
  setSelectedPrimaryShardCount,
  setSelectedReplicaCount,
  setSelectedNodeAttrs,
  setSelectedPolicyName,
} from '.';

export const fetchingIndexTemplates = createAction('FETCHING_INDEX_TEMPLATES');
export const fetchedIndexTemplates = createAction('FETCHED_INDEX_TEMPLATES');
export const fetchIndexTemplates = () => async dispatch => {
  dispatch(fetchingIndexTemplates());

  let templates;
  try {
    templates = await loadIndexTemplates();
  } catch (err) {
    return toastNotifications.addDanger(err.data.message);
  }

  dispatch(fetchedIndexTemplates(templates));
};

export const fetchedIndexTemplate = createAction('FETCHED_INDEX_TEMPLATE');
export const fetchIndexTemplate = templateName => async (dispatch) => {
  let template;
  try {
    template = await loadIndexTemplate(templateName);
  } catch (err) {
    return toastNotifications.addDanger(err.data.message);
  }

  if (template.settings && template.settings.index) {
    dispatch(
      setSelectedPrimaryShardCount(template.settings.index.number_of_shards)
    );
    dispatch(
      setSelectedReplicaCount(template.settings.index.number_of_replicas)
    );
    if (
      template.settings.index.routing &&
      template.settings.index.routing.allocation &&
      template.settings.index.routing.allocation.include
    ) {
      dispatch(
        setSelectedNodeAttrs(
          template.settings.index.routing.allocation.include.sattr_name
        )
      );
    }
    if (template.settings.index.lifecycle) {
      dispatch(setSelectedPolicyName(template.settings.index.lifecycle.name));
    }
  }

  let indexPattern = template.index_patterns[0];
  if (indexPattern.endsWith('*')) {
    indexPattern = indexPattern.slice(0, -1);
  }
  dispatch(setIndexName(`${indexPattern}-00001`));
  dispatch(setAliasName(`${indexPattern}-alias`));
  dispatch(fetchedIndexTemplate(template));
};

export const setSelectedIndexTemplateName = createAction(
  'SET_SELECTED_INDEX_TEMPLATE_NAME'
);

export const setSelectedIndexTemplate = name => async dispatch => {
  // Await all of these to ensure they happen before the next round of validation
  dispatch(setSelectedIndexTemplateName(name));
  dispatch(fetchIndexTemplate(name));
};
