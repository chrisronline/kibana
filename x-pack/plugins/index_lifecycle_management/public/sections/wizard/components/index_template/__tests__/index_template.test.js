/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import { shallow } from 'enzyme';
import { IndexTemplate } from '../index_template';
import { STRUCTURE_TEMPLATE_SELECTION, STRUCTURE_CONFIGURATION } from '../../../../../store/constants';

describe('IndexTemplate', () => {
  it('should render normally', () => {
    const component = shallow(
      <IndexTemplate
        done={() => {}}
        validate={() => {}}
        errors={{
          [STRUCTURE_TEMPLATE_SELECTION]: [],
          [STRUCTURE_CONFIGURATION]: [],
        }}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should return error state on validation', async () => {
    const validate = jest.fn();
    const component = shallow(
      <IndexTemplate
        done={() => {}}
        validate={validate}
        errors={{
          [STRUCTURE_TEMPLATE_SELECTION]: ['There is an error.'],
          [STRUCTURE_CONFIGURATION]: [],
        }}
      />
    );

    const result = await component.instance().validate();
    expect(result).toBe(false);
    expect(validate).toHaveBeenCalled();
  });

  it('should submit if valid', async () => {
    const done = jest.fn();
    const component = shallow(
      <IndexTemplate
        done={done}
        validate={() => {}}
        errors={{
          [STRUCTURE_TEMPLATE_SELECTION]: ['There is an error.'],
          [STRUCTURE_CONFIGURATION]: [],
        }}
      />
    );

    await component.instance().submit();
    expect(done).not.toHaveBeenCalled();

    const noErrorComponent = shallow(
      <IndexTemplate
        done={done}
        validate={() => {}}
        errors={{
          [STRUCTURE_TEMPLATE_SELECTION]: [],
          [STRUCTURE_CONFIGURATION]: [],
        }}
      />
    );

    await noErrorComponent.instance().submit();
    expect(done).toHaveBeenCalled();
  });
});
