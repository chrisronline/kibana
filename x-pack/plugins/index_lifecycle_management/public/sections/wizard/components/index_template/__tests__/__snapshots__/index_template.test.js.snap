// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexTemplate should render normally 1`] = `
<div
  className="euiAnimateContentLoad"
>
  <Connect(TemplateSelection)
    errors={Array []}
    isShowingErrors={false}
    validate={[Function]}
  />
  <Connect(Configuration)
    errors={Array []}
    isShowingErrors={false}
    validate={[Function]}
  />
  <EuiHorizontalRule
    className="ilmHrule"
    margin="l"
    size="full"
  />
  <EuiButton
    color="primary"
    fill={true}
    iconSide="right"
    iconType="sortRight"
    onClick={[Function]}
    type="button"
  >
    Continue
  </EuiButton>
</div>
`;
